src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.cs

```diff
@@ -0,0 +1,252 @@
+            _packageReferenceHandler.GetProjectSnapshot(selectedConfiguredProject);
+            _assemblyReferenceHandler.GetProjectSnapshot(selectedConfiguredProject);
+            _sdkReferenceHandler.GetProjectSnapshot(selectedConfiguredProject);
+        }
+
+        private List<ProjectSystemReferenceInfo> GetReferences()
+        {
+            List<ProjectSystemReferenceInfo> references = new List<ProjectSystemReferenceInfo>();
+
+            references.AddRange(_projectReferenceHandler.GetReferences());
+            references.AddRange(_packageReferenceHandler.GetReferences());
+            references.AddRange(_assemblyReferenceHandler.GetReferences());
+            references.AddRange(_sdkReferenceHandler.GetReferences());
+
+            return references;
+        }
+
+        public async Task<bool> TryUpdateReferenceAsync(string projectPath, string targetFrameworkMoniker, ProjectSystemReferenceUpdate referenceUpdate, CancellationToken cancellationToken)
+        {
+            bool wasUpdated = false;
+
+            if (referenceUpdate.Action == ProjectSystemUpdateAction.None)
+            {
+                return wasUpdated;
+            }
+
+            var activeConfiguredProject = await GetActiveConfiguredProjectByPathAsync(projectPath);
+
+            var referenceHandler = FindReferenceHandler(referenceUpdate);
+
+            wasUpdated = await ApplyAction(referenceUpdate, cancellationToken, referenceHandler, activeConfiguredProject);
+
+            return wasUpdated;
+        }
+
+        private static async Task<bool> ApplyAction(ProjectSystemReferenceUpdate referenceUpdate, CancellationToken cancellationToken,
+            ReferenceHandler referenceHandler, ConfiguredProject activeConfiguredProject)
+        {
+            bool wasUpdated;
+
+            if (referenceUpdate.Action == ProjectSystemUpdateAction.TreatAsUsed ||
+                referenceUpdate.Action == ProjectSystemUpdateAction.TreatAsUnused)
+            {
+                wasUpdated =
+                    await referenceHandler.UpdateReferenceAsync(activeConfiguredProject, referenceUpdate, cancellationToken);
+            }
+            else
+            {
+                if (!canRemoveReference(referenceUpdate, referenceHandler))
+                {
+                    wasUpdated = false;
+                }
+                else
+                {
+                    wasUpdated =
+                        await referenceHandler.RemoveReferenceAsync(activeConfiguredProject, referenceUpdate.ReferenceInfo);
+                }
+            }
+
+            return wasUpdated;
+        }
+
+        private static bool canRemoveReference(ProjectSystemReferenceUpdate referenceUpdate, ReferenceHandler referenceHandler)
+        {
+            var references = referenceHandler.GetReferences();
+
+            var reference = references.Where(c =>
+                c.ItemSpecification == referenceUpdate.ReferenceInfo.ItemSpecification).First();
```

Comment:
Ditto re `First`.

Reply:
In this case we should return False, since we cannot make changes to a reference that doesn't exist.
Done. Handled with try/catch